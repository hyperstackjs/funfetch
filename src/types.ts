export interface ParsedResponse {
  body: string | any
  status: number
  headers: Record<string, string>
}
export type SimpleResponse = string | any

export interface MethodConfig {
  bearer?: string
  basic?: { user: string; password: string }
  base?: string
  fetch?: any
  baseOpts?: RequestInit
  fullResponse?: boolean
  responseHandler?: (res: Response) => ParsedResponse | SimpleResponse
  throwIfBadStatus?: (res: Response) => void
  throwIfError?: (res: ParsedResponse | SimpleResponse) => void
}
export type Endpoint<TParams, TData, TQuery, TRes> = (
  data?: EndpointData<TParams, TData, TQuery>
) => Promise<TRes>

export type DynamicEndpoint = Endpoint<any, any, any, any>

export interface EndpointData<TParams, TData, TQuery> {
  params?: TParams | any
  data?: TData | any
  query?: TQuery | any
  opts?: RequestInit
}
export interface EndpointParams {
  ep: any
  config?: MethodConfig
}

export interface RequestParams<TParams, TData, TQuery> {
  method: string
  params?: TParams
  data?: TData
  query?: TQuery
  baseOpts?: RequestInit
  opts?: RequestInit
}

export interface Funfetch {
  post: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
  get: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
  put: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
  del: <TParams, TData, TQuery, TRes>(
    ep: string | ((params: any) => string)
  ) => Endpoint<TParams, TData, TQuery, TRes>
}
