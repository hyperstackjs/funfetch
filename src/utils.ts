export const headersToObject = (headers: Headers) => {
  const acc: Record<string, string> = {}
  for (const p of headers.entries()) {
    acc[p[0]] = p[1]
  }
  return acc
}
