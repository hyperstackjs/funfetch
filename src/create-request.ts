import isFunction from 'lodash/isFunction'
import { Headers } from 'cross-fetch'
import merge from 'lodash/merge'
import type { EndpointParams, RequestParams } from './types'
import { headersToObject } from './utils'

const buildUrl = <TQuery>(targetUrl: string, query: TQuery, base?: string) => {
  // use a dummy example.com to avoid an exception, and in order
  // to later manipulate params via a fully parsed URL
  const newUrl = new URL(targetUrl, base || 'http://example.com')
  if (query) {
    Object.entries(query).forEach(([k, v]: [string, any]) =>
      newUrl.searchParams.append(k, v)
    )
  }

  // not an absolute URL originally, no one supplied a base,
  // this means we need to strip 'example.com'
  if (!targetUrl.startsWith('http') && !base) {
    return newUrl.pathname + newUrl.search + newUrl.hash
  }
  return newUrl.toString()
}

export const createRequest = <TParams, TData, TQuery>(
  { ep, config }: EndpointParams,
  {
    method,
    params,
    data,
    query,
    baseOpts,
    opts,
  }: RequestParams<TParams, TData, TQuery>
) => {
  const headers = new Headers({
    'Content-Type': 'application/json',
  })
  if (config?.basic) {
    headers.set(
      'Authorization',
      `Basic ${btoa(`${config.basic.user}:${config.basic.password}`)}`
    )
  }
  if (config?.bearer) {
    headers.set('Authorization', `Bearer ${config.bearer}`)
  }

  const customopts: RequestInit = {}
  customopts.headers = headersToObject(headers)

  const producedUrl = isFunction(ep) ? ep(params) : ep
  const targetUrl = buildUrl(producedUrl, query, config?.base)

  return {
    url: targetUrl,
    body: data ? JSON.stringify(data) : undefined,
    method,
    opts: merge({}, baseOpts, customopts, opts),
  }
}
