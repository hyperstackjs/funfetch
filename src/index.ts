import fetch from 'cross-fetch'
import createDebug from 'debug'
import type {
  Endpoint,
  EndpointData,
  Funfetch,
  MethodConfig,
  ParsedResponse,
  SimpleResponse,
} from './types'
import { createRequest } from './create-request'
import { headersToObject } from './utils'

export * from './types'

const debug = createDebug('funfetch')

const noop = (_res: ParsedResponse | SimpleResponse) => {}

const fullResponseJsonHandler = async (
  x: Response
): Promise<ParsedResponse> => {
  const { headers } = x
  const contentType = headers.get('Content-Type')
  if (contentType?.includes('application/json')) {
    return {
      body: await x.json(),
      status: x.status,
      headers: headersToObject(x.headers),
    }
  }
  return {
    body: await x.text(),
    status: x.status,
    headers: headersToObject(x.headers),
  }
}
const simpleResponseJsonHandler = async (
  x: Response
): Promise<SimpleResponse> => {
  const { headers } = x
  const contentType = headers.get('Content-Type')
  if (contentType?.includes('application/json')) {
    return await x.json()
  }
  return await x.text()
}

const throwIfBadStatus = (response: Response) => {
  if (response.status >= 400 && response.status < 600) {
    throw new Error(JSON.stringify(response))
  }
}

export const createMethod =
  (method: string, config?: MethodConfig) =>
  <TParams, TData, TQuery, TRes>(
    ep: string | ((params: TParams | any) => string)
  ): Endpoint<TParams, TData, TQuery, TRes> =>
  async ({
    params,
    data,
    query,
    opts,
  }: EndpointData<TParams, TData, TQuery> = {}): Promise<TRes> => {
    const fetcher = config?.fetch || fetch
    const req = createRequest<TParams, TData, TQuery>(
      { ep, config },
      { method, params, data, query, opts, baseOpts: config?.baseOpts }
    )
    debug('request: %j', req)
    const res = await fetcher(req.url, {
      method: req.method,
      body: req.body,
      ...req.opts,
    })

    const throwIfBad = config?.throwIfBadStatus || throwIfBadStatus
    throwIfBad(res)

    const convertedResponse = await (
      config?.responseHandler ||
      (config?.fullResponse
        ? fullResponseJsonHandler
        : simpleResponseJsonHandler)
    )(res)

    // parsed response can be: ParsedResponse (full), string (text()), any (json())
    debug('response: %O', convertedResponse)

    const throwIfError = config?.throwIfError || noop
    throwIfError(convertedResponse)

    debug('done: (no errors)')
    return convertedResponse
  }

export const funfetch = (config: MethodConfig = {}): Funfetch => {
  return {
    post: createMethod('POST', config),
    get: createMethod('GET', config),
    put: createMethod('PUT', config),
    del: createMethod('DELETE', config),
  }
}
