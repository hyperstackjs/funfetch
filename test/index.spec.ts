import http from 'http'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import L from 'lodash'
import { funfetch } from '../src'
import app from './fixtures/app'

const normalize = (res) => {
  return L.omit(res, [
    'headers.date',
    'headers.etag',
    'headers.host',
    'body.headers.host',
  ])
}
export const client = (opts) => {
  const { post, get, del } = funfetch(opts)
  return {
    app: {
      queryString: get('/querystring'),
      queryStringDelete: del('/querystring'),
      status400WithMessage: post('/400-with-message'),
      status404: get('/404'),
      status404WithJson: get('/404-with-valid-json'),
      status404WithInvalidJson: get('/404-with-invalid-json'),
      status404WithError: get('/404-with-stripe-error'),
      delay: get('/delay'),
      users: get(({ userId }) => `/users/${userId}`),
      createUser: post(({ userId }) => `/users/${userId}`),
      postContent: post(`/post-content`),
    },
  }
}

describe('funfetch', () => {
  let server: any = null
  let serverAddr: string | null = null
  beforeAll(async () => {
    await new Promise((resolve, _reject) => {
      server = http.createServer(app as any).listen(() => {
        resolve(true)
      })
      serverAddr = `http://localhost:${server.address().port}`
    })
  })
  afterAll(() => {
    server.close()
  })
  it('full response', async () => {
    const c = client({
      base: serverAddr,
      baseOpts: {
        headers: {
          'User-Agent': 'funfetch/1.0',
        },
      },
      fullResponse: true,
      throwIfBadStatus: () => {},
    })

    expect(
      normalize(await c.app.queryString({ query: { hello: 'get query' } }))
    ).toMatchSnapshot('GET/queryString')

    expect(
      normalize(await c.app.queryStringDelete({ query: { prop: 'del query' } }))
    ).toMatchSnapshot('DELETE/queryString')

    expect(normalize(await c.app.status400WithMessage())).toMatchSnapshot(
      'POST/400 with message'
    )

    expect(normalize(await c.app.status404())).toMatchSnapshot('get/404')

    expect(normalize(await c.app.status404WithJson())).toMatchSnapshot(
      'get/404 with json'
    )

    try {
      expect(normalize(await c.app.status404WithInvalidJson())).toMatchSnapshot(
        'get/404 with invalid json'
      )
    } catch (err) {
      expect(err).toBeTruthy()
    }

    expect(normalize(await c.app.status404WithError())).toMatchSnapshot(
      'get/404 with invalid json'
    )

    expect(normalize(await c.app.delay())).toMatchSnapshot('get/delay')

    expect(
      normalize(await c.app.users({ params: { userId: 'id-22334' } }))
    ).toMatchSnapshot('get/users')

    expect(
      normalize(
        await c.app.createUser({
          params: { userId: 'id-22334' },
          data: { firstName: 'foo' },
        })
      )
    ).toMatchSnapshot('post/users')

    expect(normalize(await c.app.postContent())).toMatchSnapshot('post-content')
  })

  it('simple response: default throw status', async () => {
    const c = client({ base: serverAddr })

    try {
      await c.app.status400WithMessage()
      throw new Error('should not be here')
    } catch (e) {}
  })

  it('simple response', async () => {
    const c = client({ base: serverAddr, throwIfBadStatus: () => {} })

    expect(
      normalize(await c.app.queryString({ query: { hello: 'get query' } }))
    ).toMatchSnapshot('GET/queryString')

    expect(
      normalize(await c.app.queryStringDelete({ query: { prop: 'del query' } }))
    ).toMatchSnapshot('DELETE/queryString')

    expect(normalize(await c.app.status400WithMessage())).toMatchSnapshot(
      'POST/400 with message'
    )

    expect(normalize(await c.app.status404())).toMatchSnapshot('get/404')

    expect(normalize(await c.app.status404WithJson())).toMatchSnapshot(
      'get/404 with json'
    )

    try {
      expect(normalize(await c.app.status404WithInvalidJson())).toMatchSnapshot(
        'get/404 with invalid json'
      )
    } catch (err) {
      expect(err).toBeTruthy()
    }

    expect(normalize(await c.app.status404WithError())).toMatchSnapshot(
      'get/404 with invalid json'
    )

    expect(normalize(await c.app.delay())).toMatchSnapshot('get/delay')

    expect(
      normalize(await c.app.users({ params: { userId: 'id-22334' } }))
    ).toMatchSnapshot('get/delay')

    expect(normalize(await c.app.postContent())).toMatchSnapshot('post-content')
  })

  it('custom response handler', async () => {
    const c = client({
      base: serverAddr,
      responseHandler: (x) => x.statusText.toString(),
      throwIfBadStatus: () => {},
    })

    expect(
      await c.app.queryString({ query: { hello: 'get query' } })
    ).toMatchSnapshot('GET/queryString')
  })
})
