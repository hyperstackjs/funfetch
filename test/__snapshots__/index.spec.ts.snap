// Vitest Snapshot v1

exports[`funfetch > custom response handler > GET/queryString 1`] = `"OK"`;

exports[`funfetch > full response > DELETE/queryString 1`] = `
{
  "body": {
    "prop": "del query",
  },
  "headers": {
    "connection": "close",
    "content-length": "20",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 200,
}
`;

exports[`funfetch > full response > GET/queryString 1`] = `
{
  "body": {
    "hello": "get query",
  },
  "headers": {
    "connection": "close",
    "content-length": "21",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 200,
}
`;

exports[`funfetch > full response > POST/400 with message 1`] = `
{
  "body": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Cannot POST /400-with-message</pre>
</body>
</html>
",
  "headers": {
    "connection": "close",
    "content-length": "156",
    "content-security-policy": "default-src 'none'",
    "content-type": "text/html; charset=utf-8",
    "x-content-type-options": "nosniff",
    "x-powered-by": "Express",
  },
  "status": 404,
}
`;

exports[`funfetch > full response > get/404 1`] = `
{
  "body": "Not Found",
  "headers": {
    "connection": "close",
    "content-length": "9",
    "content-type": "text/plain; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 404,
}
`;

exports[`funfetch > full response > get/404 with invalid json 1`] = `
{
  "body": {
    "error": {
      "code": 23,
      "message": "Some error happened",
      "param": "hello_world",
      "stack": {},
    },
  },
  "headers": {
    "connection": "close",
    "content-length": "86",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 404,
}
`;

exports[`funfetch > full response > get/404 with json 1`] = `
{
  "body": {
    "foo": "baz",
  },
  "headers": {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 400,
}
`;

exports[`funfetch > full response > get/delay 1`] = `
{
  "body": {},
  "headers": {
    "connection": "close",
    "content-length": "2",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 200,
}
`;

exports[`funfetch > full response > get/users 1`] = `
{
  "body": {
    "userId": "id-22334",
  },
  "headers": {
    "connection": "close",
    "content-length": "21",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 200,
}
`;

exports[`funfetch > full response > post/users 1`] = `
{
  "body": {
    "body": {
      "firstName": "foo",
    },
    "params": {
      "userId": "id-22334",
    },
  },
  "headers": {
    "connection": "close",
    "content-length": "59",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 200,
}
`;

exports[`funfetch > full response > post-content 1`] = `
{
  "body": {
    "data": {},
    "headers": {
      "accept": "*/*",
      "accept-encoding": "gzip,deflate",
      "connection": "close",
      "content-length": "0",
      "content-type": "application/json",
      "user-agent": "funfetch/1.0",
    },
  },
  "headers": {
    "connection": "close",
    "content-length": "200",
    "content-type": "application/json; charset=utf-8",
    "x-powered-by": "Express",
  },
  "status": 200,
}
`;

exports[`funfetch > simple response > DELETE/queryString 1`] = `
{
  "prop": "del query",
}
`;

exports[`funfetch > simple response > GET/queryString 1`] = `
{
  "hello": "get query",
}
`;

exports[`funfetch > simple response > POST/400 with message 1`] = `
{
  "0": "<",
  "1": "!",
  "10": "h",
  "100": "p",
  "101": "r",
  "102": "e",
  "103": ">",
  "104": "C",
  "105": "a",
  "106": "n",
  "107": "n",
  "108": "o",
  "109": "t",
  "11": "t",
  "110": " ",
  "111": "P",
  "112": "O",
  "113": "S",
  "114": "T",
  "115": " ",
  "116": "/",
  "117": "4",
  "118": "0",
  "119": "0",
  "12": "m",
  "120": "-",
  "121": "w",
  "122": "i",
  "123": "t",
  "124": "h",
  "125": "-",
  "126": "m",
  "127": "e",
  "128": "s",
  "129": "s",
  "13": "l",
  "130": "a",
  "131": "g",
  "132": "e",
  "133": "<",
  "134": "/",
  "135": "p",
  "136": "r",
  "137": "e",
  "138": ">",
  "139": "
",
  "14": ">",
  "140": "<",
  "141": "/",
  "142": "b",
  "143": "o",
  "144": "d",
  "145": "y",
  "146": ">",
  "147": "
",
  "148": "<",
  "149": "/",
  "15": "
",
  "150": "h",
  "151": "t",
  "152": "m",
  "153": "l",
  "154": ">",
  "155": "
",
  "16": "<",
  "17": "h",
  "18": "t",
  "19": "m",
  "2": "D",
  "20": "l",
  "21": " ",
  "22": "l",
  "23": "a",
  "24": "n",
  "25": "g",
  "26": "=",
  "27": "\\"",
  "28": "e",
  "29": "n",
  "3": "O",
  "30": "\\"",
  "31": ">",
  "32": "
",
  "33": "<",
  "34": "h",
  "35": "e",
  "36": "a",
  "37": "d",
  "38": ">",
  "39": "
",
  "4": "C",
  "40": "<",
  "41": "m",
  "42": "e",
  "43": "t",
  "44": "a",
  "45": " ",
  "46": "c",
  "47": "h",
  "48": "a",
  "49": "r",
  "5": "T",
  "50": "s",
  "51": "e",
  "52": "t",
  "53": "=",
  "54": "\\"",
  "55": "u",
  "56": "t",
  "57": "f",
  "58": "-",
  "59": "8",
  "6": "Y",
  "60": "\\"",
  "61": ">",
  "62": "
",
  "63": "<",
  "64": "t",
  "65": "i",
  "66": "t",
  "67": "l",
  "68": "e",
  "69": ">",
  "7": "P",
  "70": "E",
  "71": "r",
  "72": "r",
  "73": "o",
  "74": "r",
  "75": "<",
  "76": "/",
  "77": "t",
  "78": "i",
  "79": "t",
  "8": "E",
  "80": "l",
  "81": "e",
  "82": ">",
  "83": "
",
  "84": "<",
  "85": "/",
  "86": "h",
  "87": "e",
  "88": "a",
  "89": "d",
  "9": " ",
  "90": ">",
  "91": "
",
  "92": "<",
  "93": "b",
  "94": "o",
  "95": "d",
  "96": "y",
  "97": ">",
  "98": "
",
  "99": "<",
}
`;

exports[`funfetch > simple response > get/404 1`] = `
{
  "0": "N",
  "1": "o",
  "2": "t",
  "3": " ",
  "4": "F",
  "5": "o",
  "6": "u",
  "7": "n",
  "8": "d",
}
`;

exports[`funfetch > simple response > get/404 with invalid json 1`] = `
{
  "error": {
    "code": 23,
    "message": "Some error happened",
    "param": "hello_world",
    "stack": {},
  },
}
`;

exports[`funfetch > simple response > get/404 with json 1`] = `
{
  "foo": "baz",
}
`;

exports[`funfetch > simple response > get/delay 1`] = `{}`;

exports[`funfetch > simple response > get/delay 2`] = `
{
  "userId": "id-22334",
}
`;

exports[`funfetch > simple response > post-content 1`] = `
{
  "data": {},
  "headers": {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "connection": "close",
    "content-length": "0",
    "content-type": "application/json",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
  },
}
`;
